testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala

```diff
@@ -715,9 +729,100 @@ trait ArbitraryInstances {
-  // trait.  We can move them back into the trait in the next .0.
+  implicit def arbitraryEntityDecoder[F[_], A](implicit 
+    F: Effect[F],
+    ec: TestContext,
+    g: Arbitrary[DecodeResult[F, A]]
+  ) =
+    Arbitrary(for {
+      f <- arbitrary[(Message[F], Boolean) => DecodeResult[F, A]]
+      mrs <- arbitrary[Set[MediaRange]]
+    } yield new EntityDecoder[F, A] {
+      def decode(msg: Message[F], strict: Boolean): DecodeResult[F, A] = f(msg, strict)
+      def consumes = mrs
+    })
+
+  implicit def cogenMessage[F[_]](implicit F: Effect[F], ec: TestContext): Cogen[Message[F]] =
+    Cogen[(Headers, EntityBody[F])].contramap(m => (m.headers, m.body))
+
+  implicit def cogenHeaders: Cogen[Headers] =
+    Cogen[List[Header]].contramap(_.toList)
+
+  implicit def cogenHeader: Cogen[Header] =
+    Cogen[(CaseInsensitiveString, String)].contramap(h => (h.name, h.value))
+
+  implicit def arbitraryDecodeFailure: Arbitrary[DecodeFailure] =
+    Arbitrary(oneOf(
+      genMalformedMessageBodyFailure,
+      genInvalidMessageBodyFailure,
+      genMediaTypeMissing,
+      genMediaTypeMismatch
+    ))
+
+  implicit val genMalformedMessageBodyFailure: Gen[MalformedMessageBodyFailure] =
+    for {
+      details <- arbitrary[String]
+      cause <- arbitrary[Option[Throwable]]
+    } yield MalformedMessageBodyFailure(details, cause)
+
+  implicit val genInvalidMessageBodyFailure: Gen[InvalidMessageBodyFailure] =
+    for {
+      details <- arbitrary[String]
+      cause <- arbitrary[Option[Throwable]]
+    } yield InvalidMessageBodyFailure(details, cause)
+
+  implicit val genMediaTypeMissing: Gen[MediaTypeMissing] =
+    arbitrary[Set[MediaRange]].map(MediaTypeMissing(_))
+
+  implicit val genMediaTypeMismatch: Gen[MediaTypeMismatch] =
+    for {
+      messageType <- arbitrary[MediaType]
+      expected <- arbitrary[Set[MediaRange]]
+    } yield MediaTypeMismatch(messageType, expected)
+
+  // These instances are private because they're half-baked and I don't want to encourage external use yet.
+  private[http4s] implicit val cogenDecodeFailure: Cogen[DecodeFailure] =
+    Cogen { (seed: Seed, df: DecodeFailure) =>
+      df match {
+        case MalformedMessageBodyFailure(d, t) => Cogen[(String, Option[Throwable])].perturb(seed, (d, t))
+        case InvalidMessageBodyFailure(d, t) => Cogen[(String, Option[Throwable])].perturb(seed, (d, t))
+        case MediaTypeMissing(mrs) => Cogen[Set[MediaRange]].perturb(seed, mrs)
+        case MediaTypeMismatch(mt, e) => Cogen[(MediaType, Set[MediaRange])].perturb(seed, (mt, e))
+        // TODO What if it's not one of these?
+      }
+    }
+
+  private[http4s] implicit def arbitraryMessage[F[_]]: Arbitrary[Message[F]] =
+    // TODO this is bad because the underlying generators are bad
+    Arbitrary(Gen.oneOf(arbitrary[Request[F]], arbitrary[Response[F]]))
+
+  private[http4s] implicit def arbitraryRequest[F[_]]: Arbitrary[Request[F]] = Arbitrary {
```

Comment:
This, and response, are something I've wanted for a long time, but they need to get a lot more sophisticated to make sure they're following the HTTP rules.